nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,530 @@
+
+        if (!p->configure()) {
+          return false;
+        }
+
+        polygons_.push_back(p);
+      } else if (polygon_type == "circle") {
+        std::shared_ptr<Circle> c = std::make_shared<Circle>(
+          node, polygon_name, base_frame_id, simulation_time_step);
+
+        if (!c->configure()) {
+          return false;
+        }
+
+        polygons_.push_back(c);
+      } else {  // Error if something else
+        RCLCPP_ERROR(
+          get_logger(),
+          "[%s]: Unknown polygon type: %s",
+          polygon_name.c_str() ,polygon_type.c_str());
+        return false;
+      }
+    }
+
+    // Working with data sources parameters
+    std::vector<std::string> source_names = get_parameter("observation_sources").as_string_array();
+    for (std::string source_name : source_names) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, source_name + ".type",
+        rclcpp::ParameterValue("scan"));  // Laser scanner by default
+      const std::string source_type = get_parameter(source_name + ".type").as_string();
+
+      if (source_type == "scan") {
+        std::shared_ptr<Scan> s = std::make_shared<Scan>(
+          node, tf_buffer_, source_name, base_frame_id,
+          transform_tolerance, data_timeout_);
+
+        s->configure();
+
+        sources_.push_back(s);
+      } else if (source_type == "pointcloud") {
+        std::shared_ptr<PointCloud> p = std::make_shared<PointCloud>(
+          node, tf_buffer_, source_name, base_frame_id,
+          transform_tolerance, data_timeout_);
+
+        p->configure();
+
+        sources_.push_back(p);
+      } else {  // Error if something else
+        RCLCPP_ERROR(
+          get_logger(),
+          "[%s]: Unknown source type: %s",
+          source_name.c_str(), source_type.c_str());
+        return false;
+      }
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Error while getting parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+void CollisionMonitor::processMain()
+{
+  // Current timestamp for all inner routines prolongation
+  rclcpp::Time curr_time = this->now();
+
+  // Do nothing if main worker in non-active state
+  // or if velocity is not received yet
+  if (!process_active_ || !velocityValid(curr_time)) {
+    return;
+  }
+
+  // Points array collected from different data sources in a robot base frame
+  std::vector<Point> collision_points;
+
+  // Fill collision_points array from different data sources
+  for (std::shared_ptr<SourceBase> source : sources_) {
+    source->getData(curr_time, collision_points);
+  }
+
+  // By default - there is no action
+  Action robot_action{DO_NOTHING, {0.0, 0.0, 0.0}};
+  // Used for APPROACH model to define "nearest" obstacle
+  double min_collision_time = std::numeric_limits<double>::max();
+
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    const ActionType at = polygon->getActionType();
```

Comment:
move the instantiation of `at` closer to where its used, it throw off the flow  here. Please at line 370